### ISSUE TITLE
Happy User Feedback

### ISSUE THREAD

**Issue by cruftex** [cache2k/cache2k#139]
Providing a caching library with a comprehensive set of features, consistent API and documentation is a lot of work. In case cache2k works well for you, please put a friendly comment in this issue. The feedback helps driving motivation and further development of cache2k.

Please consider the following points in your comment:

- How long are you using cache2k?
- Which features do you value the most in cache2k?
- If possible: In which products or websites do you use cache2k?
- Any random praise or critics?
- What change/improvement in cache2k would be most helpful for you?

**Comment by cr-orilibhaber** [comment]
I've came to know Cache2K only recently while looking to migrate from the long-timer Guava cache due to it's somewhat lacking support for refresh ahead loading and resilience patterns.
I needed a read-through caching solution that will be able to scale well within a multi-tenant, highly available and massively concurrent service.
After doing some research and scanning through some articles/benchmarks results, Cache2K shined as a leading alternative in terms of is modern architecture, robust out of the box feature set and superior caching algorithm implementation.
I then decided to take it for a spin and it was refreshingly joyful to find out that the praising were justified for a change.
Cache2K had proven time and time again, that the trust I put in it is well deserving!
It's _easy to use_, _fluent_, _ridiculously fast_ and has a _clean API_ that helps you focus more on you business logic without worrying about implementation details.
And as a cherry topping, it also natively supports the JCache standard that can ease the minds of all those developers who hate to commit to just one implementation 😉
I highly recommend you give it a try, you might like it... **a lot!**

**Comment by Chasewhip8** [comment]
I've personally been using cache2k for as long as I can remember, especially the documentation that is provided is super nice. I've only run into 1 area where I feel it could be improved upon but that is probably due to my use case. Overall this is a superior cache library in almost every single way and I prefer using it over anything else whenever possible.
I wouldn't recommend any other cache library if I'm being completely honest.

**Comment by vladp** [comment]
hello

1) > How long are you using cache2k?
Since early 2018

2) > Which features do you value the most in cache2k?
Java Generics, Android compatibility, JSR 107 compliance, small and does not rely on bunch of dependencies -- therefore reducing the 'surface' of potential errors and incompatibilities.

3) > If possible: In which products or websites do you use cache2k?
Cannot share product(s) a the moment. But it is used in production, low volume site.
We use it to cache data from database, as well as to cache some responses we generate.
And then if a request matches previously requested keys, we pull that from the cache. 
It helps us to dramatically reduce the load on PG server.


4) > Any random praise or critics?
Very happy with the library. Sorry I cannot verbalize better at the moment. But very happy, thank you for
making this available.

We are using the cache mostly within backend (which uses Javalin to implement REST and websockets).
Cache2K is used to cache data as well to cache responses (uncompressed JSON)


5) > What change/improvement in cache2k would be most helpful for you?

Would like to be able to set cache limit, not by number of items, but by the amount of memory heap still available (perhaps checked in last 3 minutes).
Not sure if that's feasible, but would be nice.

Would also be nice if cache2k had some sort of 'out-of-core' mode (meaning that the after say N items,
the data will be stored in high performance on-disk cache (eg. [lmdb java](1) or something else that also works with android]
At some point in time, I wanted to take a stub at it myself, but just could not figure out where even to start.

We are also planning to take a stub and see if we can integrate this with a distributed cache (infinispan).
As we need a way to make cache updates visible to distributed servers.
This is not an ask -- just mentioning this as our next expansion in the use of cache2k.

[1] https://github.com/lmdbjava/lmdbjava

**Comment by cruftex** [comment]
@vladp Thanks for the long and informative comment. That drives my motivation a lot!

> Would like to be able to set cache limit, not by number of items, but by the amount of memory heap still available

Java applications, conceptually take the amount of heap they need for proper operation. Depending on the dynamics and current operations of your application it may request more heap or free some heap. E.g. the GC might request additional heap to increase the breathing space, reduce GCs and therefore the throughput of the application when load gets higher. With caching, there is usually a minimum of items that need caching and then the applications performs acceptable. With setting these limits, the application uses the amount of heap it needs.

Using the available heap is a logical requirement, but it is reversing the concept of how Java applications operate in general. There are some approaches in that direction:

1. EHCache allows to limit caches by memory size, that makes it possible to control the memory usage a bit better. The problem here is determining the memory usage of an object correctly is tricky.
2. You can wrap each value or entry within a `WeakReference`. That means the entry is available as long as it is not garbage collected. Effectively that means you use the available heap for caching. Problems here: Memory and CPU overhead for the `WeakReference`, and the GC might a hot cache entry. There are two players now that decide which entry gets dropped: the cache and the GC.

cache2k will get a weigher in the next version, with this you can mimic setting the capacity by data size. I am also thinking about an approach how to make caching and the GC play nicer together.

> Would also be nice if cache2k had some sort of 'out-of-core' mode

Before releasing 1.0 I had done already a lot of work to support this. That is called hierarchical or layered caching. Working towards 1.0 I decided to remove this, because it appeared it was not possible to stabilize this within limited time. In the official API and features, right now, there is the `CacheLoader` and the `CacheWriter` plus some event listeners that you can use. Those are not sufficient if you want to build cache hierarchy that supports all cache operations consistently, including the atomic guarantees plus guys like `clear` and `iterate` and expiry, events etc. I am happy to share more details here. Maybe we should coordinate the work and ideas here.

**Comment by vladp** [comment]
@cruftex Thank you very much for the follow up.

WRT layered caching. 
Thank you for the offer to share/help with ideas. I wish I would have like a small prototype to share, so that you could comment on something more concrete. If I will build something like this in a couple of weekends, will bring it to your attention.
  

WRT limiting size by remaining heap -- I think I understand the challenge.
Yes may be the new weighter idea will help. 
In general, I prefer, of course, if possible, optional dynamic self-tuning for things that affect resource usage. 
To the point if there is not enough capacity to do something, slow down/throttle down but do not crash
(which is were the Layred caching would also come in).

I do not think the 'trigger' to stop adding to cache (or remove some items from there), or a trigger to switch to 'disk' caching -- needs to be perfectly precise.

A probabilistic non-linear derivative of some JVM counter or other periodically checked counter/resource -- may be good enough. 

Not sure if that's possible, but if, say, GC is slowing down/invoked more often because of memory pressure -- may be that could be one of the 'valves' that releases some items from cache2k.
In other words, not relying on individual object instance sizing, but on behaviour of the GC to see if  a 'release valve' needs to be triggered.

**Comment by cruftex** [comment]
@vladp 

Please don't spend your weekends on hacking. I spend already weeks on that problem.

Doing a generic layered caching is a very hard task. Example: Is cache.clear() expected
to free up all disk resources? What happens with concurrent reads and writes?
What happens with an ongoing iteration? While there are answers to every problem,
these have to be implemented, tested and tested with concurrence to be sure there are no
race conditions.

Distributed caching with near caching (fast local caching) is even harder.
What are the guarantees the resulting cache should have?
One guarantee/expectation is that after a value is updated, only the new value
is seen. However, that means the updater needs to wait until every peer is updated.
That, in turn, leads to different ways the cache is operated, since every operation has
more lag: more async and more bulk.

**Comment by vladp** [comment]
@cruftex  thank you for sharing  your experience. Yes indeed, both of those are too complicated, and, probably, too much of a scope creep for a clean in memory kv cache like cache2k.

**Comment by cruftex** [comment]
Closing this issue to reduce open issues that need action. I am still happy for motivating comments.
